cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/preview/PreviewErrorClassificationHttpHandler.java

```diff
@@ -99,9 +99,12 @@ public void classifyRunIdLogs(HttpRequest request, HttpResponder responder,
         LogOffset.INVALID_KAFKA_OFFSET);
     readRange = adjustReadRange(readRange, null, true);
     try (CloseableIterator<LogEvent> logIter = logReader.getLog(loggingContext,
-        readRange.getFromMillis(), readRange.getToMillis(), filter)) {
+        readRange.getFromMillis(), readRange.getToMillis(), filter);
+        CloseableIterator<LogEvent> logIter2 = logReader.getLog(loggingContext,
+            readRange.getFromMillis(), readRange.getToMillis(), filter)) {
       // the iterator is closed by the BodyProducer passed to the HttpResponder
-      errorLogsClassifier.classify(logIter, responder, namespaceId, null, "preview", previewId);
+      errorLogsClassifier.classify(logIter, logIter2,responder, namespaceId, null,
```

Comment:
Instead of passing multiple iterators, pass `Iterable<LogEvent>`.  Or even better create `inteface CloseableIterable<T> extends Iterable<T> {CloseableIterator<T> iterator();}` and use that.
It can be easily created as `CloseableIterable<LogEvent> logIter2 = () -> logReader.getLog(loggingContext, readRange.getFromMillis(), readRange.getToMillis(), filter)`. 
This way 
(a) you don't have this artificial iter1, iter2 
(b) you don't create iter2 unless needed

Reply:
Note: read my next commend about single scan first. With single scan you don't need to do it.

Reply:
done.